atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/sweep/asts/bucketingthings/DefaultSweepAssignedBucketStore.java

```diff
@@ -81,6 +97,83 @@ public BucketStateAndIdentifier getBucketStateAndIdentifier() {
                         + " initialisation, and as such, is an invalid state."));
     }
 
+    @Override
+    public Set<Bucket> getStartingBucketsForShards(Set<ShardAndStrategy> shardAndStrategies) {
+        Map<ShardAndStrategy, Cell> shardAndStrategyToCell = shardAndStrategies.stream()
+                .collect(Collectors.toMap(
+                        Function.identity(), SweepAssignedBucketStoreKeyPersister.INSTANCE::sweepBucketPointerCell));
+        Map<Cell, Long> results = readCells(shardAndStrategyToCell.values(), bucketIdentifierPersister::tryDeserialize);
+        return KeyedStream.stream(shardAndStrategyToCell)
+                .map(cell -> Optional.ofNullable(results.get(cell)))
+                .map((shardAndStrategy, bucketIdentifier) -> bucketIdentifier
+                        .map(id -> Bucket.of(shardAndStrategy, id))
+                        .orElseGet(() -> {
+                            log.warn(
+                                    "No starting bucket found for shard and strategy {}",
+                                    SafeArg.of("shardAndStrategy", shardAndStrategy));
+                            return Bucket.of(shardAndStrategy, 0);
+                        }))
+                .values()
+                .collect(Collectors.toSet());
+    }
+
+    @Override
+    // TODO: See if we should centralise the work that was done in Jeremy's PR
```

Comment:
Thoughts on centralising this "bump upwards" logic vs duplicating it just once more here? This is a total of two usages, which isn't that many. pro: more specific log lines and naming, con: need to test again (but it's not that hard to write the tests)

Given it's only two usages, a reasonable option is to deduplicate if there's yet another need for this style of logic.

Reply:
Yeah I think 3 is usually my threshold for something getting annoying enough to extract